pkg/shed/db.go

```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		db.metrics.FirstFailCount.Inc()
+	} else {
+		db.logger.Tracef("first value with prefix %s, key %s, value len %d", string(prefix), string(key), len(value))
+	}
+
+	return key, value, err
+}
+
+// Last retuns the last key matching the given prefix.
+func (db *DB) Last(prefix []byte) (key []byte, value []byte, err error) {
+	if prefix != nil {
+		db.logger.Tracef("Finding last key with prefix %s", string(prefix))
+	}
+	db.metrics.LastCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = true
+		o.PrefetchSize = 1024
+
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		for i.Seek(prefix); i.ValidForPrefix(prefix); i.Next() {
+			key = i.Item().Key()
+			value, err = i.Item().ValueCopy(value)
+			if err != nil {
+				return err
+			}
+		}
+
+		if key == nil {
+			return ErrNotFound
+		}
+
+		if !bytes.HasPrefix(key, prefix) {
+			return ErrNotFound
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while finding last value for prefix : %v", string(prefix))
+		db.logger.Debugf("error while doing last. prefix : %v, Error : %w", string(prefix), err.Error())
+		db.metrics.LastFailCount.Inc()
+	} else {
+		db.logger.Tracef("last value with prefix %s, key %s, value len %d", string(prefix), string(key), len(value))
+	}
+	return key, value, err
 }
 
-// NewIterator wraps LevelDB NewIterator method to increment metrics counter.
-func (db *DB) NewIterator() iterator.Iterator {
-	db.metrics.IteratorCounter.Inc()
-	return db.ldb.NewIterator(nil, nil)
+// GetBatch get a new badger transaction to be used for multiple atomic operations.
+func (db *DB) GetBatch(update bool) (txn *badger.Txn) {
+	db.logger.Tracef("getting a transaction with update %t", update)
+	db.metrics.GetBatchCount.Inc()
+	// set update to true indicating that data will be added/changed in this transaction.
+	return db.bdb.NewTransaction(update)
 }
 
-// WriteBatch wraps LevelDB Write method to increment metrics counter.
-func (db *DB) WriteBatch(batch *leveldb.Batch) (err error) {
-	err = db.ldb.Write(batch, nil)
+// WriteBatch commits the badger transaction after all the operations are over.
+func (db *DB) WriteBatch(txn *badger.Txn) (err error) {
+	db.metrics.WriteBatchCount.Inc()
+	err = txn.Commit()
 	if err != nil {
-		db.logger.Debugf("could not writing batch. Error: %s ", err.Error())
-		db.metrics.WriteBatchFailCounter.Inc()
+		db.logger.Errorf("error while committing transaction. Error : %w", err.Error())
+		db.logger.Debugf("error while committing batch. Error : %w", err.Error())
+		db.metrics.WriteBatchFailCount.Inc()
 		return err
 	}
-	db.metrics.WriteBatchCounter.Inc()
+	txn.Discard()
+	db.logger.Tracef("transaction committed successfully")
 	return nil
 }
 
-// Close closes LevelDB database.
-func (db *DB) Close() (err error) {
-	close(db.quit)
-	return db.ldb.Close()
+// Close shut's down the badger DB.
+func (db *DB) Close(ctx context.Context) (err error) {
```

Comment:
The context is not used. Why is it added?

Reply:
if we want to cancel Close() if it takes too long or something in future...What do you say.. keep it or remove it...?

Reply:
I do not see any reason to add an argument that is not used in the function body.

Reply:
Please, remove the unused argument.